### PR TITLE
145 node comparison speedup

### LINKED ISSUE #145: Node comparisons
# Description

Optimize and speedup node compares. Pass the comparison function along and make sure all the tests are not evaluated multiple times.

### PR DESCRIPTION
**Pull request by FBroy**
# Description

Speedup of the node comparisons.
Follow up to the value and general compare.

# Can be deleted?

Yes

# Approvals

2

### CODE REVIEW

**Review by HectorPeeters — APPROVED** [review]
Looks good to me! also good idea to add a check for ITEM when setting the AST attribute. It might be nice to centralize this check at some point though

File: src/typeInference/annotateBinaryOperator.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ export function annotateBinOp(
 	if (valueType) {
 		const type = { type: valueType, mult: left.mult };
 
-		if (valueType !== ValueType.XSNUMERIC) {
+		if (valueType !== ValueType.XSNUMERIC && valueType !== ValueType.ITEM) {
```
Comment (MTimmerije) [inline src/typeInference/annotateBinaryOperator.ts:43]: is there a particular reason for this?

**Review by MTimmerije — APPROVED** [review]

**Review by maanpalle — APPROVED** [review]
Looks good nice jobe

File: src/typeInference/annotateBinaryOperator.ts (hunk 1 above)
Comment (FBroy) [reply src/typeInference/annotateBinaryOperator.ts:43]: Yes, we don't want to annotate the AST with item()*

**Pull request merged**